src/createServer.ts

```diff
@@ -1,91 +1,84 @@
-import net = require('net')
+import http = require('http')
+import {IncomingMessage, Server, ServerResponse} from 'http'
 import {StoreController} from 'package-store'
 
 export default function (
   store: StoreController,
-  opts: object,
+  opts: {
+    path?: string,
+    port?: number,
+    hostname?: string,
+  },
 ) {
-  const server = net.createServer()
-  server.listen(opts)
-  server.on('connection', (socket) => {
-    const jsonSocket = new JsonSocket(socket)
-    const requestPackage = requestPackageWithCtx.bind(null, {jsonSocket, store})
+  const manifestPromises = {}
+  const filesPromises = {}
 
-    jsonSocket.on('message', async (message) => {
-      switch (message.action) {
-        case 'requestPackage': {
-          await requestPackage(message.msgId, message.args[0], message.args[1])
-          return
-        }
-        case 'prune': {
-          await store.prune()
-          return
-        }
-        case 'updateConnections': {
-          await store.updateConnections(message.args[0], message.args[1])
-          return
+  const server = http.createServer(async (req: IncomingMessage, res: ServerResponse) => {
+    if (req.method !== 'POST') {
+      res.statusCode = 503
+      res.end(JSON.stringify(`Only POST is allowed, received ${req.method}`))
+      return
+    }
+
+    let body: any = '' // tslint:disable-line
+    req.on('data', (data) => {
+      body += data
+    })
+    req.on('end', async () => {
+      try {
+        if (body.length > 0) {
+          body = JSON.parse(body)
+        } else {
+          body = {}
         }
-        case 'saveState': {
-          await store.saveState()
-          return
+
+        switch (req.url) {
+          case '/requestPackage':
+            const pkgResponse = await store.requestPackage(body.wantedDependency, body.options)
+            if (!pkgResponse.isLocal) {
+              manifestPromises[body.msgId] = pkgResponse.fetchingManifest
+              filesPromises[body.msgId] = pkgResponse.fetchingFiles
+            }
+            res.end(JSON.stringify(pkgResponse))
+            break
+          case '/packageFilesResponse':
+            const filesResponse = await filesPromises[body.msgId]
+            delete filesPromises[body.msgId]
+            res.end(JSON.stringify(filesResponse))
+            break
+          case '/manifestResponse':
+            const manifestResponse = await manifestPromises[body.msgId]
+            delete manifestPromises[body.msgId]
+            res.end(JSON.stringify(manifestResponse))
+            break
+          case '/updateConnections':
+            await store.updateConnections(body.prefix, body.opts)
+            res.end(JSON.stringify('OK'))
+            break
+          case '/prune':
+            await store.prune()
+            res.end(JSON.stringify('OK'))
+            break
+          case '/saveState':
+            await store.saveState()
+            res.end(JSON.stringify('OK'))
+            break
         }
```

Comment:
if no route was matched, I think a 404 response should be returned

Reply:
right 👍